File: NetFramework/DataFilter/VB/FilterEditor/FilterEditorForm.vb
Imports System.Data.OleDb
Imports System.IO
Imports System.Xml.Serialization
Imports C1.DataCollection
Imports C1.DataFilter
Imports C1.Win.C1Input
Imports C1.Win.C1Themes

Public Class FilterEditorForm
    Private Const xmlFileName As String = "FilterExpression.xml"
    Private pathToXmlFile As String = Nothing

    Private Sub FilterEditorForm_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
        FillComboBoxWithThemeNames()
        ApplyDataSource()
        ApplyFilter()
    End Sub

    Private Sub FillComboBoxWithThemeNames()
        Dim themes = C1ThemeController.GetThemes()

        For Each theme In themes
            ribbonCmbThemes.Items.Add(theme)
        Next

        ribbonCmbThemes.SelectedIndex = Array.IndexOf(themes, "Office2016Colorful")
    End Sub

    Private Sub ApplyDataSource()
        Dim dt = Data.GetDataSource()
        flexGrid.DataSource = dt
        filterEditor.DataSource = dt
    End Sub

    Private Sub ApplyFilter()
        pathToXmlFile = Directory.GetCurrentDirectory() & "\" & xmlFileName
        Dim filterExpression As CombinationExpression = Nothing

        If Not File.Exists(pathToXmlFile) Then
            filterExpression = GetPredefinedFilter()
        Else
            filterExpression = LoadFilterFromFile(pathToXmlFile)
        End If

        filterEditor.SetExpression(filterExpression)
        filterEditor.ApplyFilter()
    End Sub

    Private Shared Function GetPredefinedFilter() As CombinationExpression
        Dim filterExpression = New CombinationExpression()
        Dim filterExpressions = filterExpression.Expressions
        Dim hpGreaterThen200Expression = New OperationExpression() With {
            .PropertyName = "HP",
            .FilterOperation = FilterOperation.GreaterThanOrEqual,
            .Value = 200
        }
        filterExpressions.Add(hpGreaterThen200Expression)
        Dim automaticTransmissionExpression = New OperationExpression() With {
            .PropertyName = "AutomaticTransmission",
            .FilterOperation = FilterOperation.Equal,
            .Value = True
        }
        filterExpressions.Add(automaticTransmissionExpression)
        Dim sportsCategoryExpression = New OperationExpression() With {
            .PropertyName = "StartSaleDate",
            .FilterOperation = FilterOperation.GreaterThanOrEqual,
            .Value = New DateTime(2011, 1, 1)
        }
        filterExpressions.Add(sportsCategoryExpression)
        Dim brandOrExpression = New CombinationExpression()
        brandOrExpression.FilterCombination = FilterCombination.[Or]
        Dim brandExpressions = brandOrExpression.Expressions
        Dim brandEqualMercedesBenzExpression = New OperationExpression() With {
            .PropertyName = "Brand",
            .FilterOperation = FilterOperation.EqualText,
            .Value = "Mercedes-Benz"
        }
        brandExpressions.Add(brandEqualMercedesBenzExpression)
        Dim brandEqualAudiExpression = New OperationExpression() With {
            .PropertyName = "Brand",
            .FilterOperation = FilterOperation.EqualText,
            .Value = "Audi"
        }
        brandExpressions.Add(brandEqualAudiExpression)
        filterExpressions.Add(brandOrExpression)
        Return filterExpression
    End Function

    Private Function LoadFilterFromFile(ByVal filePath As String) As CombinationExpression
        Dim filterExpression As CombinationExpression
        Dim xmlSerializer = New XmlSerializer(GetType(CombinationExpression))

        Using fs = File.Open(filePath, FileMode.Open)
            filterExpression = TryCast(xmlSerializer.Deserialize(fs), CombinationExpression)
        End Using

        Return filterExpression
    End Function

    Private Sub ribbonCmbThemes_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles ribbonCmbThemes.SelectedIndexChanged
        themeController.Theme = ribbonCmbThemes.SelectedItem.Text
    End Sub

    Private Sub btnApply_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnApply.Click
        filterEditor.ApplyFilter()
    End Sub

    Private Sub btnReset_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnReset.Click
        filterEditor.ClearFilter()
        filterEditor.ApplyFilter()
    End Sub

    Private Sub filterEditor_FilterChanged(ByVal sender As Object, ByVal e As EventArgs) Handles filterEditor.FilterChanged
        SaveFilterToFile(pathToXmlFile)
    End Sub

    Private Sub SaveFilterToFile(ByVal filePath As String)
        Dim expression = filterEditor.GetExpression()
        Dim xmlSerializer = New XmlSerializer(GetType(CombinationExpression))

        Using fs = File.Create(filePath)
            xmlSerializer.Serialize(fs, expression)
        End Using
    End Sub

    Private Sub filterEditor_CustomEditorInitializing(sender As Object, e As C1.Win.DataFilter.CustomEditorEventArgs) Handles filterEditor.CustomEditorInitializing
        If e.PropertyName = "Brand" Then
            Dim cmb = New C1ComboBox() With {
                .AutoSize = False,
                .Width = 120
            }
            cmb.ItemsDataSource = Data.GetDataSource()
            cmb.ItemsValueMember = "Brand"
            cmb.ItemsDisplayMember = "Brand"
            e.Control = cmb
        End If
    End Sub
End Class

File: NetFramework/DataFilter/CS/FilterEditor/FilterEditorForm.cs
using C1.DataCollection;
using C1.DataFilter;
using C1.Win.C1Input;
using C1.Win.C1Themes;
using C1.Win.DataFilter;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Data.OleDb;
using System.Drawing;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using System.Xml;
using System.Xml.Serialization;

namespace FilterEditor
{
    public partial class FilterEditorForm : Form
    {
        private const string xmlFileName = "FilterExpression.xml";
        private string pathToXmlFile = null;

        public FilterEditorForm()
        {
            InitializeComponent();
            filterEditor.CustomEditorInitializing += FilterEditor_CustomEditorInitializing;
        }

        private void FilterEditorForm_Load(object sender, EventArgs e)
        {
            FillComboBoxWithThemeNames();
            ApplyDataSource();
            ApplyFilter();
        }

        private void FillComboBoxWithThemeNames()
        {
            var themes = C1ThemeController.GetThemes();
            foreach (var theme in themes)
            {
                ribbonCmbThemes.Items.Add(theme);
            }

            ribbonCmbThemes.SelectedIndex = Array.IndexOf(themes, "Office2016Colorful");
        }

        private void ApplyDataSource()
        {
            var dt = Data.GetDataSource();
            flexGrid.DataSource = dt;
            filterEditor.DataSource = dt;
        }

        private void ApplyFilter()
        {
            pathToXmlFile = Directory.GetCurrentDirectory() + "\\" + xmlFileName;
            CombinationExpression filterExpression = null;

            if (!File.Exists(pathToXmlFile))
            {
                filterExpression = GetPredefinedFilter();
            }
            else
            {
                filterExpression = LoadFilterFromFile(pathToXmlFile);
            }

            filterEditor.SetExpression(filterExpression);
            filterEditor.ApplyFilter();
        }

        private static CombinationExpression GetPredefinedFilter()
        {
            var filterExpression = new CombinationExpression();

            var filterExpressions = filterExpression.Expressions;

            var hpGreaterThen200Expression = new OperationExpression()
            {
                PropertyName = "HP",
                FilterOperation = FilterOperation.GreaterThanOrEqual,
                Value = 200
            };
            filterExpressions.Add(hpGreaterThen200Expression);

            var automaticTransmissionExpression = new OperationExpression()
            {
                PropertyName = "AutomaticTransmission",
                FilterOperation = FilterOperation.Equal,
                Value = true
            };
            filterExpressions.Add(automaticTransmissionExpression);

            var sportsCategoryExpression = new OperationExpression()
            {
                PropertyName = "StartSaleDate",
                FilterOperation = FilterOperation.GreaterThanOrEqual,
                Value = new DateTime(2011, 1, 1)
            };
            filterExpressions.Add(sportsCategoryExpression);

            var brandOrExpression = new CombinationExpression();
            brandOrExpression.FilterCombination = FilterCombination.Or;

            var brandExpressions = brandOrExpression.Expressions;

            var brandEqualMercedesBenzExpression = new OperationExpression()
            {
                PropertyName = "Brand",
                FilterOperation = FilterOperation.EqualText,
                Value = "Mercedes-Benz"
            };
            brandExpressions.Add(brandEqualMercedesBenzExpression);

            var brandEqualAudiExpression = new OperationExpression()
            {
                PropertyName = "Brand",
                FilterOperation = FilterOperation.EqualText,
                Value = "Audi"
            };
            brandExpressions.Add(brandEqualAudiExpression);

            filterExpressions.Add(brandOrExpression);

            return filterExpression;
        }

        private CombinationExpression LoadFilterFromFile(string filePath)
        {
            CombinationExpression filterExpression;
            var xmlSerializer = new XmlSerializer(typeof(CombinationExpression));
            using (var fs = File.Open(filePath, FileMode.Open))
            {
                filterExpression = xmlSerializer.Deserialize(fs) as CombinationExpression;
            }

            return filterExpression;
        }

        private void ribbonCmbThemes_SelectedIndexChanged(object sender, EventArgs e)
        {
            themeController.Theme = ribbonCmbThemes.SelectedItem.Text;
        }

        private void btnApply_Click(object sender, EventArgs e)
        {
            filterEditor.ApplyFilter();
        }

        private void btnReset_Click(object sender, EventArgs e)
        {
            filterEditor.ClearFilter();
            filterEditor.ApplyFilter();
        }

        private void filterEditor_FilterChanged(object sender, EventArgs e)
        {
            SaveFilterToFile(pathToXmlFile);
        }

        private void SaveFilterToFile(string filePath)
        {
            var expression = filterEditor.GetExpression();
            var xmlSerializer = new XmlSerializer(typeof(CombinationExpression));
            using (var fs = File.Create(filePath))
            {
                xmlSerializer.Serialize(fs, expression);
            }
        }

        private void FilterEditor_CustomEditorInitializing(object sender, CustomEditorEventArgs e)
        {
            if (e.PropertyName == "Brand")
            {
                var cmb = new C1ComboBox() { AutoSize = false, Width = 120 };
                cmb.ItemsDataSource = Data.GetDataSource();
                cmb.ItemsValueMember = "Brand";
                cmb.ItemsDisplayMember = "Brand";
                e.Control = cmb;                
            }
        }
    }
}
